packages/core/src/operator/operator-runtime/getWinningKeyCount.ts

```diff
@@ -0,0 +1,77 @@
+import { ethers } from "ethers";
+
+/**
+ * @dev Calculates the number of winning keys for a bulk submission, this is a off-chain implementation of the same function inside the Referee
+ * @param _keyCount Amount of keys for the bulk submission (# of keys staked in pool, or # of unstaked keys for an owner)
+ * @param _boostFactor The factor controlling the chance of eligibility for payout as a multiplicator
+ * @param _bulkAddress The Address to be used for randomize
+ * @param _challengeId The ID of the challenge. Is only used for randomize
+ * @param _confirmData The confirm data of the assertion.
+ * @param _challengerSignedHash The signed hash for the challenge
+ * @return {number} winningKeyCount Number of winning keys
+ */
+export const getWinningKeyCount = (keyCount: number, boostFactor: number, bulkAddress: string, challengeId: bigint, confirmData: string, challengerSignedHash: string): number => {
+    if (keyCount == 0) {
+        return 0;
+    }
+
+    const probability = boostFactor;
+
+    // Create a unique seed for this submission
+    const seedHash = ethers.keccak256(
+        ethers.solidityPacked(
+            ['address', 'uint256', 'bytes', 'bytes'],
+            [bulkAddress, challengeId, confirmData, challengerSignedHash]
+        )
+    );
+    const seed = BigInt(seedHash);
+
+    const expectedWinningKeys = Math.floor((keyCount * probability) / 10000);
+
+    if (expectedWinningKeys === 0) {
+        const scaleFactor = 1000000;
+        const scaledProbability = Math.floor((probability * scaleFactor) / 10000);
+        const scaledExpectedWinningKeys = BigInt(keyCount) * BigInt(scaledProbability);
+
+        const randomThreshold = generateRandomNumber(seed, "threshold") % BigInt(scaleFactor);
+
+        return randomThreshold < scaledExpectedWinningKeys ? 1 : 0;
+    }
+
+    if(expectedWinningKeys > 0 && expectedWinningKeys < 4) {
+        const randomThreshold = generateRandomNumber(seed, "threshold") % BigInt(1000);
+        const randomFactor = generateRandomNumber(seed, "factor") % BigInt(2);
+        
+        if(randomThreshold < 300){
+            if(randomFactor === BigInt(0)){
+                return expectedWinningKeys + 1;
+            }else{
+                return expectedWinningKeys - 1;
+            }
+        }else{
+            return expectedWinningKeys;
+        }
+    }
+
+    const baseVariability = 30 + Math.floor(1000 / boostFactor);
+    const maxAdjustmentPercentage = Math.min(baseVariability, 50);
```

Comment:
I believe this should be `const maxAdjustmentPercentage = Math.min(baseVariability, 30)`

Reply:
Updated to maxAdjustment `30%`

@XDapps I got this from current `fb-tiny-keys` `infrastructure\smart-contracts\test\utils\getWinningKeyCountLocal.mjs`

It seems to be `50` there too, does that need to be updated as well ?
And did we run the test report with the "wrong" value ?

Reply:
Actually yes and no. 

After review, we need to adjust the function slightly.

This => `uint256 baseVariability = 30 + (1000 / _boostFactor)` is irrelevant as it will return a number greater than 30 every time.

This => `uint256 maxAdjustmentPercentage = baseVariability > 30
            ? 30
            : baseVariability`
            
  Can be changed to `uint256 maxAdjustmentPercentage = 30`
  
  The current implementation is returning 30 as the max adjustmentPercentage for all scenarios, we're now going to remove that redundant calculation and just set the maxAdjustmentPercentage to 30. 
  
  Summary: the report stats would not have changed, but I am happy to re-run them if you'd like.
  
  Let me know if you want me to make a PR to update that function in the contract and local test.

Reply:
I understand, I have updated the getWInningKeyCount in this PR accordingly